subsys/bluetooth/controller/ll_sw/ull_llcp.c

```diff
@@ -90,18 +84,115 @@ void llcp_proc_ctx_release(struct proc_ctx *ctx)
 	mem_release(ctx, &mem_ctx.free);
 }
 
-bool llcp_tx_alloc_is_available(void)
+#if (CONFIG_BT_CTLR_LLCP_TX_BUFFER_QUEUE_SIZE > 0)
+/*
+ * @brief Check for per conn pre-allocated tx buffer allowance
+ * @return true if buffer is available
+ */
+static inline bool static_tx_buffer_available(struct proc_ctx *ctx)
+{
+#if (CONFIG_BT_CTLR_LLCP_PER_CONN_TX_CTRL_BUF_NUM > 0)
+	/* Check if per connection pre-aloted tx buffer is available */
+	if (ctx->conn->llcp.tx_buffer_alloc < CONFIG_BT_CTLR_LLCP_PER_CONN_TX_CTRL_BUF_NUM) {
+		/* This connection has not yet used up all the pre-aloted tx buffers */
+		return true;
+	}
+#endif /* CONFIG_BT_CTLR_LLCP_PER_CONN_TX_CTRL_BUF_NUM > 0 */
+	return false;
+}
+
+/*
+ * @brief pre-alloc/peek of a tx buffer, leave requester on the wait list (@head if first up)
+ *
+ * @return true if alloc is allowed, false if not
+ *
+ */
+bool llcp_tx_alloc_peek(struct proc_ctx *ctx)
+{
+	if (!static_tx_buffer_available(ctx)) {
+		/* The conn already has spent its pre-aloted tx buffer(s),
+		 * so we should consider the common tx buffer pool
+		 */
+		if (ctx->wait_reason == WAITING_FOR_NOTHING) {
+			/* The current procedure is not in line for a tx buffer
+			 * so sign up on the wait list
+			 */
+			sys_slist_append(&tx_buffer_wait_list, &ctx->wait_node);
+			ctx->wait_reason = WAITING_FOR_TX_BUFFER;
+		}
+
+		/* Now check to see if this procedure context is @ head of the wait list */
+		if (ctx->wait_reason == WAITING_FOR_TX_BUFFER &&
+		    sys_slist_peek_head(&tx_buffer_wait_list) == &ctx->wait_node) {
+			return (common_tx_buffer_alloc <
+				CONFIG_BT_CTLR_LLCP_COMMON_TX_CTRL_BUF_NUM);
+		}
+
+		return false;
+	}
+	return true;
+}
+
+/*
+ * @brief un-peek of a tx buffer, in case ongoing alloc is aborted
+ *
+ */
+void llcp_tx_alloc_unpeek(struct proc_ctx *ctx)
+{
+	sys_slist_remove(&tx_buffer_wait_list, NULL, &ctx->wait_node);
```

Comment:
Can we always be sure that unpeek is done on a head node as the NULL indicates? Would `sys_slist_find_and_remove()` be better?

Reply:
No, we cannot - I had understood the remove as doing find and remove, when passing NULL as previous node. I will replace!